Add stunnel package for pfSense 2.3

djmarcin opened the pull request:
Mostly straight copies of the files from the 2.2 package with the new 2.3 package metadata files.

No attempt has been made to clean up the existing code, notably in security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
- There is code checking for pfsense versions < 2.3 which can probably be deleted
- The gui attempts to render colored text, but since it gets escaped simply renders text like &lt;font color="#AABBCC"&gt;blahblah&lt;/font&gt;.  I couldn't figure out how to get colored text in the status page tables.

I am willing to work on those issues, but since this doesn't seem to have any major issues perhaps it is OK as a first approximation to get the stunnel package working again.  If the pull can't be accepted as is, any pointers on how to fix the above issues would be greatly appreciated.  I couldn't find anything promising in the convert to bootstrap docs.

rbgarga commented on security/pfSense-pkg-stunnel/Makefile:
```diff
@@ -0,0 +1,47 @@
+MAINTAINER=	coreteam@pfsense.org
+COMMENT=	pfSense package stunnel
+
+LICENSE=	ESF
```
License changed to APACHE20

rbgarga commented on security/pfSense-pkg-stunnel/files/etc/inc/priv/stunnel.priv.inc:
```diff
@@ -0,0 +1,39 @@
+	INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+	CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+	ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+	POSSIBILITY OF SUCH DAMAGE.
```
License has changed to APACHE20, replace text accordingly. You can see an example at:
https://github.com/pfsense/FreeBSD-ports/blob/devel/security/pfSense-pkg-sudo/files/etc/inc/priv/sudo.priv.inc#L2

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
```diff
@@ -0,0 +1,250 @@
+	INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+	CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+	ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+	POSSIBILITY OF SUCH DAMAGE.
```
Fix license text

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
```diff
@@ -0,0 +1,250 @@
+	define('STUNNEL_LOCALBASE', '/usr/pbi/stunnel-' . php_uname("m"));
+} else {
+	define('STUNNEL_LOCALBASE', '/usr/local');
+}
```
We don't need to care about pfSense 2.1 or 2.2 here, this conditional can be removed

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
```diff
@@ -0,0 +1,250 @@
+}
+define('STUNNEL_ETCDIR', STUNNEL_LOCALBASE . "/etc/stunnel");
+if (!isset($_GET['id']) and !isset($_POST['id'])) {
+	if ($GLOBALS['config']['installedpackages']['stunnelcerts']['savemsg']) {
```
Remove all use of $GLOBALS. The standard way in pfSense is to declare 'global $config' and access $config direct

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.xml:
```diff
@@ -0,0 +1,174 @@
+	INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+	CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+	ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+	POSSIBILITY OF SUCH DAMAGE.
```
Replace license text

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel_certs.xml:
```diff
@@ -0,0 +1,143 @@
+	INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+	CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+	ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+	POSSIBILITY OF SUCH DAMAGE.
```
Replace license text

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel_certs.xml:
```diff
@@ -0,0 +1,143 @@
+	]]>
+	</copyright>
+	<name>stunnelcerts</name>
+	<version>5.20.2</version>
```
Replace 5.20.2 by macro %%PKGVERSION%%

rbgarga commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.xml:
```diff
@@ -0,0 +1,174 @@
+	]]>
+	</copyright>
+	<name>stunnel</name>
+	<version>5.20.3</version>
```
Replace 5.20.3 by %%PKGVERSION%%

rbgarga commented on security/pfSense-pkg-stunnel/Makefile:
```diff
@@ -0,0 +1,47 @@
+	${INSTALL_DATA} ${FILESDIR}${DATADIR}/info.xml \
+		${STAGEDIR}${DATADIR}
+	@${REINPLACE_CMD} -i '' -e "s|%%PKGVERSION%%|${PKGVERSION}|" \
+		${STAGEDIR}${DATADIR}/info.xml
```
Add XML files that contain macro %%PKGVERSION%% to this REINPLACE_CMD statement

rbgarga commented on security/pfSense-pkg-stunnel/Makefile:
```diff
@@ -0,0 +1,47 @@
+
+PORTNAME=	pfSense-pkg-stunnel
+PORTVERSION=	5.30
+PORTREVISION=	1
```
New packages shouldn't contain PORTREVISION line, please remove

doktornotor commented:
I fixed most of the review comments (plus additional things unnoticed by @rbgarga) in https://github.com/djmarcin/FreeBSD-ports/pull/1

I have not touched any of the $GLOBALS stuff since I think the page should use certs.inc and store certs for the package in System - Cert Manager, rather than reimplementing it.

djmarcin commented:
I merged in the changes from @doktornotor

rbgarga commented:
As suggested by @doktornotor, change it to use certs.inc instead of reimplementing it

djmarcin commented:
I'll probably be able to look at this over the weekend, but to be honest
I'm not terribly familiar with the certs part of stunnel, I'm using it
without certs.

On Wed, Dec 21, 2016 at 11:07 AM Renato Botelho <notifications@github.com>
wrote:

> *@rbgarga* requested changes on this pull request.
>
> As suggested by @doktornotor <https://github.com/doktornotor>, change it
> to use certs.inc instead of reimplementing it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pfsense/FreeBSD-ports/pull/135#pullrequestreview-14028471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHQF8VJGdwL-KjEs1Kw2rzXTPQOotyvks5rKXgMgaJpZM4ImPOn>
> .
>

doktornotor commented:
@djmarcin - Need some help here?

djmarcin commented:
This fell off my radar, but I looked at it last night. To clarify the goal
here, we want to eliminate the certs page in the stunnel config UI and just
allow the user to choose from the system cert manager storage?

doktornotor commented:
Yeah, exactly, the XML can just let users pick up certificates for each tunnel from those set up in the Cert Manager, see e.g. https://github.com/pfsense/FreeBSD-ports/blob/devel/www/pfSense-pkg-squid/files/usr/local/pkg/squid_reverse_general.xml

Plus if you need code to do something with the certificates, use certs.inc and the functions in there if possible.

djmarcin commented:
Ok -- I've removed the custom certs implementation and replaced it with certs.inc functions.  As far as I can tell, stunnel needs access to .pem files in order to function, and I couldn't figure out if the system cert manager provided access to such files -- I think not? So I pull the referenced certs on save and write out .pem files in the stunnel directory and delete them when they're no longer in use by stunnel.

As to whether or not stunnel actually works correctly with these certs, I *think* so but I haven't tested it rigorously.  Will give it some more testing tomorrow.

doktornotor commented:
Looks much better. (And yeah, you need to write out the certificates somewhere to filesystem, nothing wrong with that.)

djmarcin commented:
Finally got a chance to do that testing.  Verified that stunnel using a Let's Encrypt cert imported to cert manager successfully acted in server mode and provided SSL access to pfsense being served over the HTTP port, so it looks like this is working properly.

Forgive my ignorance, do I need to do anything to mark the requested changes done?

rbgarga commented on security/pfSense-pkg-stunnel/Makefile:
```diff
@@ -0,0 +1,45 @@
+	${INSTALL_DATA} ${FILESDIR}/etc/inc/priv/stunnel.priv.inc \
+		${STAGEDIR}/etc/inc/priv
+	${INSTALL_DATA} ${FILESDIR}${PREFIX}/www/shortcuts/pkg_stunnel.inc \
+                ${STAGEDIR}${PREFIX}/www/shortcuts
```
Use TAB instead of spaces here, otherwise portlint will complain

djmarcin replied:
Done (responding so it gets marked complete)

rbgarga commented on security/pfSense-pkg-stunnel/pkg-plist:
```diff
@@ -0,0 +1,6 @@
+www/shortcuts/pkg_stunnel.inc
+/etc/inc/priv/stunnel.priv.inc
+%%DATADIR%%/info.xml
+@dir /etc/inc/priv
```
It's missing '@dir /etc/inc' item here

djmarcin replied:
Done (responding so it gets marked complete)

djmarcin commented:
Done.

I noticed something while doing this.  The config file is being clobbered on installation by a version without tunnel definitions.  I attempted to fix this by replacing the deleted lines in c9d58d9 with stunnel_save() but it still didn't work.  The pfSense UI still shows the tunnels, and after saving any tunnel, all the tunnels get written out again, however it doesn't seem like the script is aware of this config during stunnel_install().

Is there another function I can implement that would be called after install but have access to the config stored by pfSense?

doktornotor commented:
@djmarcin You mean like putting

```
<custom_php_resync_config_command>
	stunnel_save();
</custom_php_resync_config_command>
```

to the XML, or what is exactly is the issue?

jim-p commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
```diff
@@ -0,0 +1,133 @@
+}
+
+function stunnel_save($config) {
+	conf_mount_rw();
```
`conf_mount_rw()` and `conf_mount_ro()` calls can be removed. It is a no-op on 2.3 and irrelevant on 2.4

jim-p commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
```diff
@@ -0,0 +1,133 @@
+function stunnel_install() {
+	safe_mkdir(STUNNEL_ETCDIR);
+
+	// Generate a self-signed default certificate.
```
You should check here to make sure such a cert does not already exist, otherwise any time the package is upgraded or reinstalled, it will make another similar certificate.

djmarcin replied:
done

jim-p commented on security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc:
```diff
@@ -0,0 +1,133 @@
+		}
+	}
+
+	conf_mount_ro();
```
`conf_mount_rw()` and `conf_mount_ro()` calls can be removed. It is a no-op on 2.3 and irrelevant on 2.4

rbgarga replied:
It's a noop on 2.4 but still used on 2.3 since 2.3 still supports nanobsd. I suggest to keep them for now so this package can be cherry-picked to 2.3

djmarcin replied:
kept for now

djmarcin commented:
There's just one problem with this which is that when the package is upgraded (well, with pkg install) it doesn't seem to start the service.  It will only start the service after some edit is made.

djmarcin commented:
It appears there is a bug with either pfSense or stunnel where start_service does not work in the context of 'pkg install' or 'pkg upgrade'.  The bug causes the service not to daemonize and become adopted by init -- instead the processes is adopted by the pkg command and then killed when that exits.  See my post in https://forum.pfsense.org/index.php?topic=111563.0 for some details.

The service is properly started on boot or on config changes though.

Added two minor cleanup edits to the branch.  Let me know if you want me to squash these or anything.

rbgarga commented:
@djmarcin package looks OK now but I believe squashing all these commits is a good idea and it'll make it easy to add it to 2.3.3 when it's properly tested for some time. Can you do it please?

djmarcin commented:
Sorry for the delay, commits have been squashed to one commit, attribution for the changes by doktornotor via "Co-authored-by" in this commit.
